### DISCUSSION TITLE
How to transcode/convert existing Flacs to 16bit / 44100Hz while importing?

### DISCUSSION THREAD

**Discussion by piedrok** [beetbox/beets#5401] [Q&A] [answered]
Hi! 

In my collection I have some huge Flac files with different frequencies and sample rates. 

Now I'd like to convert (or transcode) these to 16bit / 44100Hz Flac files. 

The command: 

`ffmpeg -i $source -y -sample_fmt s16 -ar 44100 -dither_method triangular $dest` 

does the job. 

But while the "beet convert" doesn't convert Flacs to Flacs... instead (so it says in the docs) it just copies the file. 

How can I change this behaviour to not ignore existing Flac files while importing? 

Ideally only ignoring Flac files that already are in the 16bit / 44100Hz fomat... 

Thanks for reading and maybe some help!
Cheers, p.

**Comment by Serene-Arc** [comment, marked as answer]
The convert plugin does exactly what you're asking! I'm not sure which part of the docs you read but you can definitely convert FLACs and use your own custom ffmpeg command. Here's an example configuration for the plugin.

```yaml
convert:
  format: flac
  max_bitrate: ????
  formats:
    flac:
      command: ffmpeg -i $source -y -sample_fmt s16 -ar 44100 -dither_method triangular $dest
      extension: flac
```

The operative option here is `max_bitrate`. Copying directly from the docs:

```max_bitrate: By default, the plugin does not transcode files that are already in the destination format. This option instead also transcodes files with high bitrates, even if they are already in the same format as the output. Note that this does not guarantee that all converted files will have a lower bitrate—that depends on the encoder and its configuration. Default: none.```

So set this to a number that you don't want songs to go above, and it will convert everything that goes above it with the given FLAC command. You can also run the FLAC command manually but this should do it automatically.

As always, test this on a test library and group of files to make sure that it won't ruin your music and behaves how you'd like.

**Reply by piedrok** [reply]
Thanks for your explanation! 

I already set up the convert plugin in that way - but I do not know about the bitrate ... 

Maybe I got the wrong vocabulary here (not an native speaker!). 

How can I force the conversion of files with a sample rate of 48000Hz to 44100Hz with an encoding depth of only 16bit? 

Isn't their a similar option for this as "max_bitrate"? 

Or am I missing something here? 

Thx, p.

**Comment by RollingStar** [comment]
Be cautious. To reduce the chance of artifacts, people recommend downsampling to a common factor.

192 kHz or 96 kHz converts to 48 kHz
176.4 kHz or 88.2 kHz converts to 44.1 kHz

SoX allegedly does not produce artifacts. I don't know if ffmpeg produces artifacts. https://sourceforge.net/projects/sox/

**Reply by Serene-Arc** [reply]
This is something important that I missed. You should always downsample to a common factor @piedrok. Even SoX **can** produce artifacts doing this because it's just how maths works with digital files. This will almost certainly creates strange things in at least a large portion of your files. I sort of assumed that you were trying to get 24-bit files to 16-bit, which is fine, but re-encoding files like this can cause issues.

**Reply by piedrok** [reply]
Thx everybody. 

Now I am using sox )to avoid artefacts) and I've set the "max_bitrate: 1280"... 

Basically all 16bit / 44100HZ files just get copied and I can use them as they are. 
All files with higher bit depth and/or a higher sample rate usually show bitrates over 1280  - so they get transcoded and dithered (once!). 

I've understood that there might be some artefacts in some cases still but to be honest I haven't encountered them yet - as least to my ears the comparison of a few 192kHz files and their respective 44100Hz versions doesn't show any audible differences...  

I guess this is a good compromise for me... :-) 

So thanks everybody! - Cheers! 

BTW: Is there a query to list all tracks that have been formatted to play above a certain sample rate, bit depth or bit rate within my collection?

**Reply by Serene-Arc** [reply]
The bitrate thing is only 'usually' because of how FLAC files work. Technically, if you have 10 minutes of silence and one minute of 24-bit 192 kHz music, the bitrate will be really low, because that silence is compressed down to nothing and bitrate is an average. That's not how music actually works, but it's why the compression means that it doesn't matter so much what the bitdepth is; bit rate is generally the factor you care about.

You can definitely search for those! You can use `beet list samplerate:192000` for example to get all files with a 192 kHz sample rate, or `beet list bitrate:1200..` to get all files with a bitrate above 1200. You can pass them into the convert command as well, and it'll convert them all to your desired format.

**Reply by piedrok** [reply]
Thank you very much for your explanations ! 

Now I will search my tracks for the different bitrates and have a look at their bit rates. If they seem to have way too big file sizes for my taste I might transcode them... - if not I leave them in place. 

Cheers!

**Reply by Serene-Arc** [reply]
You can use `beet info` from the info plugin to display all of the information for a track, or list just the bitrates for a query with `beet info -i bitrate`. Enjoy!

**Reply by piedrok** [reply]
Hi once more! 

The bitrate info query proved really useful - thx again. 

But I cannot figure out - sorry to bother you again - how to filter the results to only show all files with a bitrate above 1280 kb (just an example...)... is there a way to do that?  

Basically instead of listing thousands of bitrates for individual tracks I'd like to get a short list of only those tracks that exceed my tolerated maximum bitrate... 

My workaround would be to re-encode my whole library with the max_bitrate setting you suggested (see above). In theory every file already in the correct (well, my chosen...) format should be omitted and only the higher than max bitrate files would be re-encoded as batch. 

But I rather get a list first to check if that's really the way to go... 

Do you have any idea how to filter the info output in that way? 

Thank you for reading! 
Cheers, p.

**Reply by RollingStar** [reply]
Seems like you should use beet list + custom list output. there should be docs about it.

> `beet list bitrate:1200..` to get all files with a bitrate above 1200. You can pass them into the convert command as well, and it'll convert them all to your desired format.

The `..` is a greater than or less than query. The format could have the name and bitrate, or whatever fields you need.

**Reply by piedrok** [reply]
Thank you so very much! 

The formats from the docs seem to be a bit unintuitive - well, but it'll do... - also good that these commands can be combined - I'll figure it out... :-) - it's exactly what I've been looking for! 

Thx again, best regards, 
pk
